### PR TITLE
Update ex10

### PR DESCRIPTION
**Pull request by laefrost**

### CODE REVIEW

File: exercises/R/R-lab-10-template.Rmd [hunk 1]
```diff
@@ -271,3 +271,168 @@ plot(hist)
...
+
+`<!--#!solutionend-->`
+
+
+One problem of the traditional self-attention mechanism is the memory consumption of $\mathcal{O}(n^2)$, due to stroing the full $N \times N$ matrices. An approach that mitigates this bottleneck is FlashAttention. The core idea of FlashAttention is to divide the matrices $Q, K, V$ into blocks along the first dimension and perform the softmax operation block by block, s.t. you iterativley update the softmax-values of the previous blocks, with the values obtained for the current block. For example, let’s consider a specific query row $q_i \in \mathbb{R}^{1 \times d_k}$ from $Q$, and a blockwise decomposition of the key matrix $K$ into two blocks, $K^{(1)} \in \mathbb{R}^{b_1 \times d_k}$ and $K^{(2)} \in \mathbb{R}^{b_2 \times d_k}$, corresponding to the first and second chunks of the key matrix along the sequence dimension. The corresponding query-key dot products yield the two row vectors $\mathbf{a}_i^{(1)} = q_i {K^{(1)}}^\top$ and $\mathbf{a}_i^{(2)} = q_i {K^{(2)}}^\top$. In the end, we want to compute the softmax over the fully concatenated vector $\mathbf{a}_i = \left(\mathbf{a}_i^{(1)}, \mathbf{a}_i^{(2)}\right)$.
```
Comment (EmanuelSommer) [inline exercises/R/R-lab-10-template.Rmd:333]: Typo: _storing_

It is just one $N \times N$ matrix not multiple (singular)

File: exercises/R/R-lab-10-template.Rmd [hunk 2]
```diff
@@ -271,3 +271,168 @@ plot(hist)
...
+which is exactly the vanilla softmax.
+`<!--#!solutionend-->`
+
+
+## Exercise 4
+
+Now, in the last exercise we want to apply the self attention concept to out model from exercise 1. Reusing both the data preprocessing functions and the training loop, we only have to update the model:
+
+```{r}
+model <- keras_model_sequential() |>
```
Comment (EmanuelSommer) [inline exercises/R/R-lab-10-template.Rmd:404]: remove dublicated lstm model here

File: exercises/python/templates/py-lab-10-template.md [hunk 3]
```diff
@@ -918,3 +918,355 @@ def plot_prediction(model: nn.Module, dataset: Dataset, idx: int) -> None:
...
+
+#!TAG HWEND
+
+<!-- #region -->
+One problem of the traditional self-attention mechansim, is the memory consumption of $\mathcal{O}(n^2)$, due to stroing the full $N \times N$ matrices. An approach that mitigates this bottleneck is FlashAttention. The core idea of FlashAttention is to divide the matrices $Q, K, V$ into blocks along the first dimension and perform the softmax operation block by block, s.t. you iterativley update the softmax-values of the previous blocks, with the values obtained for the current block. For example, let’s consider a specific query row $q_i \in \mathbb{R}^{1 \times d_k}$ from $Q$, and a blockwise decomposition of the key matrix $K$ into two blocks, $K^{(1)} \in \mathbb{R}^{b_1 \times d_k}$ and $K^{(2)} \in \mathbb{R}^{b_2 \times d_k}$, corresponding to the first and second chunks of the key matrix along the sequence dimension. The corresponding query-key dot products yield the two row vectors $\mathbf{a}_i^{(1)} = q_i {K^{(1)}}^\top$ and $\mathbf{a}_i^{(2)} = q_i {K^{(2)}}^\top$. In the end, we want to compute the softmax over the fully concatenated vector $\mathbf{a}_i = \left(\mathbf{a}_i^{(1)}, \mathbf{a}_i^{(2)}\right)$.
```
Comment (EmanuelSommer) [inline exercises/python/templates/py-lab-10-template.md:981]: same typos as above pls fix

File: exercises/python/templates/py-lab-10-template.md [hunk 4]
```diff
@@ -918,3 +918,355 @@ def plot_prediction(model: nn.Module, dataset: Dataset, idx: int) -> None:
...
+            #!TAG HWEND
+        )
+
+        self.fc = nn.Linear(in_features=64, out_features=1)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x: Tensor) -> Tensor:
+        #!TAG HWBEGIN
+        #!MSG TODO: Implement the forward pass.
+        # The output needs to be reshaped or otherwise we have a dimension too much.
```
Comment (EmanuelSommer) [inline exercises/python/templates/py-lab-10-template.md:1088]: add a hint for the dimension squeezing to the TODO

**Review by EmanuelSommer — CHANGES_REQUESTED** [review]
Thanks! Only some minor typos to fix, then the PR is good to go :+1:
